File: isHappy.js
var isHappy = function(n) {
  let obj = {};
  while (true) {
    let sum = 0;
    let nums = n.toString().split('').map(el => Number(el) ** 2).forEach(el => sum += el);
    if (obj[sum]) {
      return false;
    }
    if (sum === 1) {
      return true;
    }
    obj[sum] = true;
    n = sum;
  }

};

console.log(isHappy(19));

File: is_happy.rb
# iterative
# def is_happy(n)
#     hash = {}
#     while true
#       sum = 0
#       nums = n.to_s.chars.map {|el| el.to_i}.each {|el| sum += el ** 2}
#       return false if hash[sum]
#       return true if sum == 1
#       hash[sum] = true
#       n = sum
#     end
# end

# recursive
def is_happy(n)
  @arr = []
  sum = 0
  n.to_s.chars.map {|el| el.to_i}.each {|el| sum += el ** 2}
  return false if @arr.include?(sum)
  return true if sum == 1
  @arr << sum
  is_happy(sum)
end


p is_happy(19)
